akka-persistence-query/src/main/scala/akka/persistence/query/journal/leveldb/DeliveryBuffer.scala

```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright (C) 2009-2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.persistence.query.journal.leveldb
+
+import akka.stream.actor.ActorPublisher
+
+trait DeliveryBuffer[T] { _: ActorPublisher[T] ⇒
+
+  var buf = Vector.empty[T]
+
+  def deliverBuf(): Unit =
+    if (buf.nonEmpty && totalDemand > 0) {
+      if (buf.size == 1) {
+        // optimize for this common case
+        onNext(buf.head)
+        buf = Vector.empty
+      } else if (totalDemand <= Int.MaxValue) {
+        val (use, keep) = buf.splitAt(totalDemand.toInt)
+        buf = keep
+        use foreach onNext
+      } else {
+        buf foreach onNext
+        buf = Vector.empty
+      }
+    }
```

Comment:
Oh yeah. I actually think this is useful enough to put somewhere in streams... I've re-implemented `deliverBuf` far too many times by now